src/compiler.rs

```diff
@@ -143,20 +149,26 @@ pub fn get_nast(
     Ok((nast, new_node_id))
 }
 
-pub fn compile(
+pub fn compile<B: Backend>(
     sources: &Sources,
-    tast: TypeChecker,
-    double_generic_gate_optimization: bool,
-) -> miette::Result<CompiledCircuit> {
-    CircuitWriter::generate_circuit(tast, double_generic_gate_optimization).into_miette(sources)
+    tast: TypeChecker<B>,
+    backend: B,
+) -> miette::Result<CompiledCircuit<B>> {
+    CircuitWriter::generate_circuit(tast, backend).into_miette(sources)
+}
+
+pub struct GeneratedWitness<B: Backend> {
+    pub all_witness: Witness<B::Field>,
+    pub full_public_inputs: Vec<B::Field>,
+    pub public_outputs: Vec<B::Field>,
```

Comment:
need documentation

Reply:
* all_witness is the entire witness vector, comprising the public input and the private input
* full_public_inputs is the public input part of the previous vector, including the public output at the end
* public_outputs is only the public output part of the previous vector

to be honest I don't remember why I returned that :)

Reply:
(seems kinda inefficient)